java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java

```diff
@@ -305,6 +321,64 @@ public void testWriteReadWithDictionaries() throws IOException {
     }
   }
 
+  @Test
+  public void testWriteReadWithStructDictionaries() throws IOException {
+    DictionaryProvider.MapDictionaryProvider provider = new DictionaryProvider.MapDictionaryProvider();
+    provider.put(dictionary4);
+
+    final StructVector vector = newVector(StructVector.class, "D4", MinorType.STRUCT, allocator);
+    final Map<String, List<Integer>> values = new HashMap<>();
+    // Index:                     0, 2, 1, 2, 1, 0, 0
+    values.put("a", Arrays.asList(1, 3, 2, 3, 2, 1, 1));
+    values.put("b", Arrays.asList(4, 6, 5, 6, 5, 4, 4));
+    setVector(vector, values);
+    FieldVector encodedVector = (FieldVector) DictionaryEncoder.encode(vector, dictionary4);
+
+    List<Field> fields = Arrays.asList(encodedVector.getField());
+    List<FieldVector> vectors = Collections2.asImmutableList(encodedVector);
+    try (VectorSchemaRoot root = new VectorSchemaRoot(fields, vectors, encodedVector.getValueCount());
+         ByteArrayOutputStream out = new ByteArrayOutputStream();
+         ArrowFileWriter writer = new ArrowFileWriter(root, provider, newChannel(out));) {
+
+      writer.start();
+      writer.writeBatch();
+      writer.end();
+
+      try (SeekableReadChannel channel = new SeekableReadChannel(
+          new ByteArrayReadableSeekableByteChannel(out.toByteArray()));
+          ArrowFileReader reader = new ArrowFileReader(channel, allocator)) {
+        final VectorSchemaRoot readRoot = reader.getVectorSchemaRoot();
+        final Schema readSchema = readRoot.getSchema();
+        assertEquals(root.getSchema(), readSchema);
+        assertEquals(1, reader.getDictionaryBlocks().size());
+        assertEquals(1, reader.getRecordBlocks().size());
+
+        reader.loadNextBatch();
+        assertEquals(1, readRoot.getFieldVectors().size());
+        assertEquals(1, reader.getDictionaryVectors().size());
+        
+        final FieldVector readEncoded = readRoot.getVector(0);
+        assertTrue(readEncoded instanceof IntVector);
```

Comment:
should we validate `readEncoded`?

Reply:
Yes. It is interesting to check the `Field`. However, using the `RangeEqualsVisitor` we can check both without any more code. So I did it.